api/tools/src/main/java/org/pmiops/workbench/tools/BackfillBillingProjectUsers.java

```diff
@@ -0,0 +1,166 @@
+
+  private static ApiClient newApiClient(String apiUrl) throws IOException {
+    ApiClient apiClient = new ApiClient();
+    apiClient.setBasePath(apiUrl);
+    GoogleCredential credential =
+        GoogleCredential.getApplicationDefault().createScoped(Arrays.asList(FC_SCOPES));
+    credential.refreshToken();
+    apiClient.setAccessToken(credential.getAccessToken());
+    return apiClient;
+  }
+
+  private static WorkspacesApi newWorkspacesApi(String apiUrl) throws IOException {
+    WorkspacesApi api = new WorkspacesApi();
+    api.setApiClient(newApiClient(apiUrl));
+    return api;
+  }
+
+  private static BillingApi newBillingApi(String apiUrl) throws IOException {
+    BillingApi api = new BillingApi();
+    api.setApiClient(newApiClient(apiUrl));
+    return api;
+  }
+
+  private static void dryLog(boolean dryRun, String msg) {
+    String prefix = "";
+    if (dryRun) {
+      prefix = "[DRY RUN] Would have... ";
+    }
+    log.info(prefix + msg);
+  }
+
+  /**
+   * Swagger Java codegen does not handle the WorkspaceACL model correctly; it returns a GSON map
+   * instead. Run this through a typed Gson conversion process to coerce it into the desired type.
+   */
+  private static Map<String, WorkspaceAccessEntry> extractAclResponse(WorkspaceACL aclResp) {
+    Type accessEntryType = new TypeToken<Map<String, WorkspaceAccessEntry>>() {}.getType();
+    Gson gson = new Gson();
+    return gson.fromJson(gson.toJson(aclResp.getAcl(), accessEntryType), accessEntryType);
+  }
+
+  private static void backfill(
+      WorkspacesApi workspacesApi,
+      BillingApi billingApi,
+      String billingProjectPrefix,
+      boolean dryRun)
+      throws ApiException {
+    int userUpgrades = 0;
+    for (WorkspaceResponse resp : workspacesApi.listWorkspaces()) {
+      Workspace w = resp.getWorkspace();
+      if (!w.getNamespace().startsWith(billingProjectPrefix)) {
+        continue;
+      }
+
+      String id = w.getNamespace() + "/" + w.getName();
+      if (!"PROJECT_OWNER".equals(resp.getAccessLevel())) {
+        log.warning(
+            String.format(
+                "service account has '%s' access to workspace '%s'; skipping",
+                resp.getAccessLevel(), id));
+        continue;
+      }
+
+      Map<String, WorkspaceAccessEntry> acl =
+          extractAclResponse(workspacesApi.getWorkspaceAcl(w.getNamespace(), w.getName()));
+      for (String user : acl.keySet()) {
+        WorkspaceAccessEntry entry = acl.get(user);
+        if (!"OWNER".equals(entry.getAccessLevel())) {
+          // Only owners should be granted billing project user.
+          continue;
+        }
+        dryLog(
+            dryRun,
+            String.format(
+                "granting billing project user on '%s' to '%s' (%s)",
+                w.getNamespace(), user, entry.getAccessLevel()));
+        if (!dryRun) {
+          billingApi.addUserToBillingProject(w.getNamespace(), "user", user);
```

Comment:
Should we be handling error cases or does this tool have a high probability on passing on every run?

One possible error is calling this API for a user that no longer exists in FC/Google.

Also, I'm guessing the call is idempotent but it looks like we would be running the same api calls if the tool is rerun.

Reply:
Good call - warning is better here, added.

This call should be idempotent. It may or may not have converged on rerun depending on the issue/ FC implementation.